File: legacy/src/render/markdown.ts
/**
 * The the reference design Markdown renderer: markdown text → styled terminal lines.
 *
 * This is the successor to the transformer from the extension era. Owning the
 * frontend removes every constraint that made that code contorted: no escape
 * games to survive re-parsing, no gutter guessing, no hard-break smuggling —
 * blocks render straight to their final lines at a known width.
 *
 * Visual contract, pinned by tests against the reference design's Zig test literals:
 *   headings   level-specific SGR (bold+underline / bold / underline / …)
 *   bullets    dim `•`, two columns of indent per level, numbers kept
 *   code       shrink-wrapped `┌ lang ─┐` panel, the reference design highlight colors
 *   quotes     dim `│ ` rail, body upright
 *   rules      fixed 60 columns, SGR dim
 *   inline     bold/italic/strike as SGR; code in the reference design's inline-code gray;
 *              links underline-only with OSC 8 hyperlink
 *
 * Parsing uses marked's lexer (re-exported by pi-tui), the same library the engine's
 * own renderer is built on.
 */

import { Marked, visibleWidth, type Tokens } from "@earendil-works/pi-tui";
import {
	BOLD_ON,
	DIM_ON,
	UNDERLINE_OFF,
	UNDERLINE_ON,
	WEIGHT_OFF,
	bold,
	dim,
	headingStyle,
	rule,
} from "./ansi.ts";
import { highlight } from "./style.ts";
import { theme } from "./style.ts";

const marked = new Marked();

const ITALIC_ON = "\x1b[3m";
const ITALIC_OFF = "\x1b[23m";
const STRIKE_ON = "\x1b[9m";
const STRIKE_OFF = "\x1b[29m";
const OSC8 = (url: string) => `\x1b]8;;${url}\x1b\\`;
const OSC8_CLOSE = "\x1b]8;;\x1b\\";

/** the reference design's inline-code color: a dedicated constant, not the statusline gray. */
function inlineCode(text: string): string {
	return theme().fg("dim", text);
}

/** Render inline tokens to a single styled string. */
export function renderInline(tokens: ReadonlyArray<Tokens.Generic> | undefined): string {
	if (!tokens) return "";
	let out = "";
	for (const token of tokens) {
		switch (token.type) {
			case "text": {
				const t = token as Tokens.Text;
				out += t.tokens ? renderInline(t.tokens) : t.text;
				break;
			}
			case "escape":
				out += (token as Tokens.Escape).text;
				break;
			case "strong":
				out += `${BOLD_ON}${renderInline((token as Tokens.Strong).tokens)}${WEIGHT_OFF}`;
				break;
			case "em":
				out += `${ITALIC_ON}${renderInline((token as Tokens.Em).tokens)}${ITALIC_OFF}`;
				break;
			case "del":
				out += `${STRIKE_ON}${renderInline((token as Tokens.Del).tokens)}${STRIKE_OFF}`;
				break;
			case "codespan":
				out += inlineCode((token as Tokens.Codespan).text);
				break;
			case "link": {
				// the reference design: underline only, no color, OSC 8 hyperlink, no printed URL.
				const l = token as Tokens.Link;
				const label = renderInline(l.tokens) || l.text;
				out += `${OSC8(l.href)}${UNDERLINE_ON}${label}${UNDERLINE_OFF}${OSC8_CLOSE}`;
				break;
			}
			case "image":
				out += (token as Tokens.Image).text;
				break;
			case "br":
				out += "\n";
				break;
			case "html":
				out += (token as Tokens.HTML).raw;
				break;
			default:
				out += "raw" in token ? String(token.raw) : "";
		}
	}
	return out;
}

const ANSI_RUN = /\x1b\[[0-9;]*m|\x1b\][^\x07\x1b]*(?:\x07|\x1b\\)/g;

function plainWidth(styled: string): number {
	return visibleWidth(styled.replace(ANSI_RUN, ""));
}

/**
 * Word-wrap a styled string. ANSI runs travel with the word they precede;
 * a row never splits a word unless the word alone exceeds the width.
 */
export function wrapStyled(styled: string, width: number): string[] {
	const rows: string[] = [];
	for (const hardLine of styled.split("\n")) {
		const words = hardLine.split(" ");
		let row = "";
		let rowWidth = 0;
		for (const word of words) {
			const w = plainWidth(word);
			if (row !== "" && rowWidth + 1 + w > width) {
				rows.push(row);
				row = word;
				rowWidth = w;
			} else {
				row = row === "" ? word : `${row} ${word}`;
				rowWidth = row === word ? w : rowWidth + 1 + w;
			}
		}
		rows.push(row);
	}
	return rows.length > 0 ? rows : [""];
}

/**
 * the reference design's code panel — geometry from `codePanelWidth`/`appendCodePanelHeader` in
 * the reference design's `transcript_blocks.zig`, pinned byte-exact by the ported tests.
 */
export function codePanel(code: string, language: string, cols: number): string[] {
	const source = code.replace(/\n+$/, "");
	const lines = language ? safeHighlight(source, language) : source.split("\n");

	const maxCodeWidth = lines.reduce((max, l) => Math.max(max, plainWidth(l)), 0);
	const labelWidth = language === "" ? 0 : Math.min(visibleWidth(language), Math.max(0, cols - 5));
	const panelWidth = Math.min(cols, Math.max(6, Math.max(maxCodeWidth + 4, labelWidth + 5)));
	const innerWidth = panelWidth - 4;

	const out: string[] = [];
	if (labelWidth > 0) {
		const label = language.slice(0, panelWidth - 5) || "?";
		const edge = "─".repeat(Math.max(0, panelWidth - 4 - visibleWidth(label)));
		out.push(`┌ ${DIM_ON}${label}${WEIGHT_OFF} ${edge}┐`);
	} else {
		out.push(`┌${"─".repeat(panelWidth - 2)}┐`);
	}
	for (const line of lines) {
		for (const row of hardWrap(line, innerWidth)) {
			const pad = " ".repeat(Math.max(0, innerWidth - plainWidth(row)));
			out.push(`│ ${row}${pad} │`);
		}
	}
	out.push(`└${"─".repeat(panelWidth - 2)}┘`);
	return out;
}

function safeHighlight(source: string, language: string): string[] {
	try {
		const lines = highlight(source, language);
		return lines.length > 0 ? lines.map(stripNoopFg) : source.split("\n");
	} catch {
		return source.split("\n");
	}
}

/**
 * the engine's highlighter styles every token; ours maps several token kinds to the
 * default color, which leaves no-op `\x1b[39m…\x1b[39m` pairs behind. the reference design emits
 * nothing for unstyled tokens, and the panel tests are byte-exact, so drop any
 * default-fg reset that has no color open to reset.
 */
function stripNoopFg(line: string): string {
	let open = false;
	let out = "";
	for (const token of line.split(/(\x1b\[[0-9;]*m)/).filter(Boolean)) {
		if (token === "\x1b[39m") {
			if (open) {
				out += token;
				open = false;
			}
			continue;
		}
		if (/^\x1b\[38[;0-9]*m$/.test(token)) open = true;
		out += token;
	}
	return out;
}

const RESETS = new Set(["\x1b[0m", "\x1b[39m", "\x1b[22m", "\x1b[24m"]);

/** Hard-wrap one code line, closing and reopening any open color at the seam. */
function hardWrap(line: string, width: number): string[] {
	if (plainWidth(line) <= width) return [line];
	const rows: string[] = [];
	let open: string | null = null;
	let row = "";
	let rowWidth = 0;
	for (const token of line.split(/(\x1b\[[0-9;]*m)/).filter(Boolean)) {
		if (token.startsWith("\x1b")) {
			open = RESETS.has(token) ? null : token;
			row += token;
			continue;
		}
		for (const ch of token) {
			const w = visibleWidth(ch);
			if (rowWidth + w > width) {
				rows.push(open ? `${row}\x1b[0m` : row);
				row = open ?? "";
				rowWidth = 0;
			}
			row += ch;
			rowWidth += w;
		}
	}
	if (rowWidth > 0 || rows.length === 0) rows.push(open ? `${row}\x1b[0m` : row);
	return rows;
}

function renderList(token: Tokens.List, depth: number, width: number): string[] {
	const out: string[] = [];
	const pad = "  ".repeat(depth);
	let n = typeof token.start === "number" && token.start > 0 ? token.start : 1;
	for (const item of token.items) {
		const glyph = token.ordered ? `${n}. ` : `${dim("•")}${" "}`;
		const glyphWidth = token.ordered ? `${n}. `.length : 2;
		const hanging = pad + " ".repeat(glyphWidth);
		const bodyWidth = Math.max(8, width - pad.length - glyphWidth);

		// An item's tokens are inline content followed by any nested blocks.
		let firstLine = true;
		for (const child of item.tokens ?? []) {
			if (child.type === "list") {
				out.push(...renderList(child as Tokens.List, depth + 1, width));
				continue;
			}
			const inline =
				child.type === "text"
					? renderInline((child as Tokens.Text).tokens ?? [child])
					: renderInline([child]);
			for (const row of wrapStyled(inline, bodyWidth)) {
				out.push(firstLine ? `${pad}${glyph}${row}` : `${hanging}${row}`);
				firstLine = false;
			}
		}
		n++;
	}
	return out;
}

function renderTable(token: Tokens.Table, width: number): string[] {
	// the reference design's table glyphs: ` │ ` separators, `─┼─` junctions.
	const header = token.header.map((c) => renderInline(c.tokens));
	const rows = token.rows.map((r) => r.map((c) => renderInline(c.tokens)));
	const cols = header.length;
	const widths = header.map((h, i) =>
		Math.max(plainWidth(h), ...rows.map((r) => plainWidth(r[i] ?? ""))),
	);
	const line = (cells: string[], boldRow: boolean) =>
		cells
			.map((c, i) => {
				const padded = c + " ".repeat(Math.max(0, widths[i]! - plainWidth(c)));
				return boldRow ? bold(padded) : padded;
			})
			.join(` ${dim("│")} `);
	const out = [line(header, true)];
	out.push(dim(widths.map((w) => "─".repeat(w)).join("─┼─")));
	for (const r of rows) out.push(line(r.length === cols ? r : [...r, ...Array(cols - r.length).fill("")], false));
	return out.map((l) => (plainWidth(l) > width ? l : l));
}

/** Render a full markdown document to lines at `width`. Blocks separated by one blank row. */
export function renderMarkdown(markdown: string, width: number): string[] {
	const out: string[] = [];
	const push = (lines: string[]) => {
		if (out.length > 0 && lines.length > 0) out.push("");
		out.push(...lines);
	};

	for (const token of marked.lexer(markdown)) {
		switch (token.type) {
			case "space":
				break;
			case "heading":
				push([headingStyle((token as Tokens.Heading).depth, renderInline((token as Tokens.Heading).tokens))]);
				break;
			case "paragraph":
				push(wrapStyled(renderInline((token as Tokens.Paragraph).tokens), width));
				break;
			case "code":
				push(codePanel((token as Tokens.Code).text, ((token as Tokens.Code).lang ?? "").trim(), width));
				break;
			case "list":
				push(renderList(token as Tokens.List, 0, width));
				break;
			case "blockquote": {
				const inner = renderMarkdown((token as Tokens.Blockquote).text, Math.max(8, width - 2));
				push(inner.map((l) => `${DIM_ON}│ ${WEIGHT_OFF}${l}`));
				break;
			}
			case "hr":
				push([rule()]);
				break;
			case "table":
				push(renderTable(token as Tokens.Table, width));
				break;
			case "html":
			case "text":
			default:
				push(wrapStyled("raw" in token ? String(token.raw).trimEnd() : "", width));
		}
	}
	return out;
}

File: crates/tui/src/render/markdown.rs
//! Markdown → styled terminal lines, in the reference design's shapes.
//!
//! Visual contract, byte-pinned by the parity tests:
//!   headings   level-specific SGR (see `ansi::heading_style`)
//!   bullets    dim `•`, two columns of indent per level, numbers kept
//!   code       shrink-wrapped `┌ lang ─┐` panel, palette highlight colors
//!   quotes     dim `│ ` rail, body upright
//!   rules      fixed 60 columns, SGR dim
//!   inline     bold/italic/strike as SGR; code spans in the palette's
//!              inline-code gray; links underline-only with OSC 8
//!
//! Parsing uses pulldown-cmark; rendering owns the width, so blocks land on
//! their final lines directly.

use pulldown_cmark::{Event, HeadingLevel, Options, Parser, Tag, TagEnd};
use unicode_width::UnicodeWidthChar;

use crate::render::ansi::*;
use crate::render::highlight::highlight_line;
use crate::render::theme::Theme;

fn osc8(url: &str) -> String {
    format!("\x1b]8;;{url}\x1b\\")
}
const OSC8_CLOSE: &str = "\x1b]8;;\x1b\\";

/// Visible width of a styled string (ANSI SGR and OSC sequences are zero).
pub fn visible_width(styled: &str) -> usize {
    let mut width = 0;
    let mut chars = styled.chars().peekable();
    while let Some(c) = chars.next() {
        if c == '\x1b' {
            match chars.peek() {
                Some('[') => {
                    while let Some(&n) = chars.peek() {
                        chars.next();
                        if n.is_ascii_alphabetic() { break; }
                    }
                }
                Some(']') => {
                    // OSC … terminated by BEL or ST (ESC \)
                    while let Some(n) = chars.next() {
                        if n == '\x07' { break; }
                        if n == '\x1b' { chars.next(); break; }
                    }
                }
                _ => {}
            }
            continue;
        }
        width += c.width().unwrap_or(0);
    }
    width
}

/// Word-wrap a styled string; ANSI runs travel with the word they precede.
pub fn wrap_styled(styled: &str, width: usize) -> Vec<String> {
    let mut rows = Vec::new();
    for hard in styled.split('\n') {
        let mut row = String::new();
        let mut row_width = 0usize;
        for word in hard.split(' ') {
            let w = visible_width(word);
            if !row.is_empty() && row_width + 1 + w > width {
                rows.push(std::mem::take(&mut row));
                row_width = 0;
            }
            if row.is_empty() {
                row.push_str(word);
                row_width = w;
            } else {
                row.push(' ');
                row.push_str(word);
                row_width += 1 + w;
            }
        }
        rows.push(row);
    }
    if rows.is_empty() { rows.push(String::new()); }
    rows
}

const RESETS: &[&str] = &["\x1b[0m", "\x1b[39m", "\x1b[22m", "\x1b[24m"];

/// Hard-wrap one code line, closing and reopening any open color at the seam.
fn hard_wrap(line: &str, width: usize) -> Vec<String> {
    if visible_width(line) <= width {
        return vec![line.to_string()];
    }
    let mut rows = Vec::new();
    let mut open: Option<String> = None;
    let mut row = String::new();
    let mut row_width = 0usize;
    let mut chars = line.chars().peekable();
    while let Some(c) = chars.next() {
        if c == '\x1b' && chars.peek() == Some(&'[') {
            let mut seq = String::from("\x1b");
            while let Some(&n) = chars.peek() {
                seq.push(n);
                chars.next();
                if n.is_ascii_alphabetic() { break; }
            }
            open = if RESETS.contains(&seq.as_str()) { None } else { Some(seq.clone()) };
            row.push_str(&seq);
            continue;
        }
        let w = c.width().unwrap_or(0);
        if row_width + w > width {
            if open.is_some() { row.push_str("\x1b[0m"); }
            rows.push(std::mem::take(&mut row));
            if let Some(o) = &open { row.push_str(o); }
            row_width = 0;
        }
        row.push(c);
        row_width += w;
    }
    if row_width > 0 || rows.is_empty() {
        if open.is_some() { row.push_str("\x1b[0m"); }
        rows.push(row);
    }
    rows
}

/// The shrink-wrapped code panel; geometry byte-pinned by the parity tests.
pub fn code_panel(theme: &Theme, code: &str, language: &str, cols: usize) -> Vec<String> {
    let source = code.trim_end_matches('\n');
    let lines: Vec<String> = if language.is_empty() {
        source.split('\n').map(String::from).collect()
    } else {
        source.split('\n').map(|l| highlight_line(theme, language, l)).collect()
    };

    let max_code_width = lines.iter().map(|l| visible_width(l)).max().unwrap_or(0);
    let label_width = if language.is_empty() { 0 } else { language.chars().count().min(cols.saturating_sub(5)) };
    let panel_width = (max_code_width + 4).max(label_width + 5).max(6).min(cols);
    let inner_width = panel_width - 4;

    let mut out = Vec::new();
    if label_width > 0 {
        let label: String = language.chars().take(panel_width.saturating_sub(5)).collect();
        let label = if label.is_empty() { "?".to_string() } else { label };
        let edge = "─".repeat(panel_width.saturating_sub(4 + label.chars().count()));
        out.push(format!("┌ {DIM_ON}{label}{WEIGHT_OFF} {edge}┐"));
    } else {
        out.push(format!("┌{}┐", "─".repeat(panel_width - 2)));
    }
    for line in &lines {
        for row in hard_wrap(line, inner_width) {
            let pad = " ".repeat(inner_width.saturating_sub(visible_width(&row)));
            out.push(format!("│ {row}{pad} │"));
        }
    }
    out.push(format!("└{}┘", "─".repeat(panel_width - 2)));
    out
}

#[derive(Default)]
struct ListState {
    ordered: Option<u64>,
}

/// Emit the current item's inline text as glyph-prefixed, hanging-indented rows.
fn flush_item(rows: &mut Vec<String>, lists: &mut [ListState], inline: &mut String, width: usize) {
    let depth = lists.len().saturating_sub(1);
    let pad = "  ".repeat(depth);
    let Some(state) = lists.last_mut() else { return };
    let (glyph, glyph_width) = match &mut state.ordered {
        Some(n) => {
            let g = format!("{n}. ");
            let w = g.chars().count();
            *n += 1;
            (g, w)
        }
        None => (format!("{}{}", dim("•"), " "), 2),
    };
    let hanging = format!("{pad}{}", " ".repeat(glyph_width));
    let body_width = width.saturating_sub(pad.chars().count() + glyph_width).max(8);
    for (i, row) in wrap_styled(inline.trim_end(), body_width).into_iter().enumerate() {
        if i == 0 { rows.push(format!("{pad}{glyph}{row}")); }
        else { rows.push(format!("{hanging}{row}")); }
    }
    inline.clear();
}

/// Render a markdown document to lines at `width`, one blank row between blocks.
pub fn render_markdown(theme: &Theme, markdown: &str, width: usize) -> Vec<String> {
    let mut opts = Options::empty();
    opts.insert(Options::ENABLE_STRIKETHROUGH);
    opts.insert(Options::ENABLE_TABLES);
    let parser = Parser::new_ext(markdown, opts);

    let mut out: Vec<String> = Vec::new();
    let push_block = |out: &mut Vec<String>, lines: Vec<String>| {
        if lines.is_empty() { return; }
        if !out.is_empty() { out.push(String::new()); }
        out.extend(lines);
    };

    // Inline accumulation state.
    let mut inline = String::new();
    let mut heading: Option<u8> = None;
    let mut lists: Vec<ListState> = Vec::new();
    let mut item_first_lines: Vec<String> = Vec::new(); // rendered rows of current list block
    // One flag per open item: has its own inline text been emitted yet?
    let mut item_stack: Vec<bool> = Vec::new();
    let mut quote_depth = 0usize;
    let mut code: Option<(String, String)> = None; // (lang, buffer)
    let mut table: Option<(Vec<String>, Vec<Vec<String>>, bool)> = None; // header, rows, in_header

    for event in parser {
        match event {
            Event::Start(Tag::Heading { level, .. }) => {
                heading = Some(match level {
                    HeadingLevel::H1 => 1, HeadingLevel::H2 => 2, HeadingLevel::H3 => 3,
                    HeadingLevel::H4 => 4, HeadingLevel::H5 => 5, HeadingLevel::H6 => 6,
                });
                inline.clear();
            }
            Event::End(TagEnd::Heading(_)) => {
                let text = heading_style(heading.take().unwrap_or(2), &inline);
                push_block(&mut out, vec![text]);
                inline.clear();
            }
            Event::Start(Tag::Paragraph) => inline.clear(),
            Event::End(TagEnd::Paragraph) => {
                if !item_stack.is_empty() {
                    // handled at item end via `inline`
                } else if quote_depth > 0 {
                    let body_width = width.saturating_sub(2).max(8);
                    let rail = quote_rail();
                    let rows: Vec<String> =
                        wrap_styled(&inline, body_width).into_iter().map(|r| format!("{rail}{r}")).collect();
                    push_block(&mut out, rows);
                    inline.clear();
                } else if table.is_none() {
                    push_block(&mut out, wrap_styled(&inline, width));
                    inline.clear();
                }
            }
            Event::Start(Tag::BlockQuote(_)) => { quote_depth += 1; }
            Event::End(TagEnd::BlockQuote(_)) => { quote_depth = quote_depth.saturating_sub(1); }
            Event::Start(Tag::List(start)) => {
                // A list opening inside an item means the item's own text is
                // done — emit it now so children render below their parent.
                if let Some(flushed) = item_stack.last_mut() {
                    if !*flushed {
                        flush_item(&mut item_first_lines, &mut lists, &mut inline, width);
                        *flushed = true;
                    }
                }
                lists.push(ListState { ordered: start });
                if lists.len() == 1 { item_first_lines.clear(); }
            }
            Event::End(TagEnd::List(_)) => {
                lists.pop();
                if lists.is_empty() {
                    push_block(&mut out, std::mem::take(&mut item_first_lines));
                }
            }
            Event::Start(Tag::Item) => { item_stack.push(false); inline.clear(); }
            Event::End(TagEnd::Item) => {
                let flushed = item_stack.pop().unwrap_or(false);
                if !flushed {
                    flush_item(&mut item_first_lines, &mut lists, &mut inline, width);
                }
                inline.clear();
            }
            Event::Start(Tag::CodeBlock(kind)) => {
                let lang = match kind {
                    pulldown_cmark::CodeBlockKind::Fenced(l) => l.split_whitespace().next().unwrap_or("").to_string(),
                    _ => String::new(),
                };
                code = Some((lang, String::new()));
            }
            Event::End(TagEnd::CodeBlock) => {
                if let Some((lang, buffer)) = code.take() {
                    push_block(&mut out, code_panel(theme, &buffer, &lang, width));
                }
            }
            Event::Start(Tag::Table(_)) => { table = Some((Vec::new(), Vec::new(), false)); }
            Event::Start(Tag::TableHead) => { if let Some(t) = &mut table { t.2 = true; } }
            Event::End(TagEnd::TableHead) => { if let Some(t) = &mut table { t.2 = false; } }
            Event::Start(Tag::TableRow) => { if let Some(t) = &mut table { if !t.2 { t.1.push(Vec::new()); } } }
            Event::Start(Tag::TableCell) => inline.clear(),
            Event::End(TagEnd::TableCell) => {
                if let Some((header, rows, in_header)) = &mut table {
                    if *in_header { header.push(std::mem::take(&mut inline)); }
                    else if let Some(last) = rows.last_mut() { last.push(std::mem::take(&mut inline)); }
                }
            }
            Event::End(TagEnd::Table) => {
                if let Some((header, rows, _)) = table.take() {
                    push_block(&mut out, render_table(&header, &rows));
                }
            }
            Event::Rule => push_block(&mut out, vec![rule()]),
            Event::Start(Tag::Strong) => inline.push_str(BOLD_ON),
            Event::End(TagEnd::Strong) => inline.push_str(WEIGHT_OFF),
            Event::Start(Tag::Emphasis) => inline.push_str(ITALIC_ON),
            Event::End(TagEnd::Emphasis) => inline.push_str(ITALIC_OFF),
            Event::Start(Tag::Strikethrough) => inline.push_str(STRIKE_ON),
            Event::End(TagEnd::Strikethrough) => inline.push_str(STRIKE_OFF),
            Event::Start(Tag::Link { dest_url, .. }) => {
                inline.push_str(&osc8(&dest_url));
                inline.push_str(UNDERLINE_ON);
            }
            Event::End(TagEnd::Link) => {
                inline.push_str(UNDERLINE_OFF);
                inline.push_str(OSC8_CLOSE);
            }
            Event::Code(text) => inline.push_str(&theme.fg("mdCode", &text)),
            Event::Text(text) => {
                if let Some((_, buffer)) = &mut code { buffer.push_str(&text); }
                else { inline.push_str(&text); }
            }
            Event::SoftBreak => inline.push(' '),
            Event::HardBreak => inline.push('\n'),
            Event::Html(html) | Event::InlineHtml(html) => inline.push_str(&html),
            _ => {}
        }
    }
    out
}

/// Table: ` │ ` separators, dim `─┼─` junction row, bold header.
fn render_table(header: &[String], rows: &[Vec<String>]) -> Vec<String> {
    let cols = header.len();
    let mut widths: Vec<usize> = header.iter().map(|h| visible_width(h)).collect();
    for row in rows {
        for (i, cell) in row.iter().enumerate() {
            if i < cols { widths[i] = widths[i].max(visible_width(cell)); }
        }
    }
    let line = |cells: &[String], bold_row: bool| -> String {
        cells.iter().enumerate().map(|(i, c)| {
            let pad = " ".repeat(widths.get(i).copied().unwrap_or(0).saturating_sub(visible_width(c)));
            let cell = format!("{c}{pad}");
            if bold_row { bold(&cell) } else { cell }
        }).collect::<Vec<_>>().join(&format!(" {} ", dim("│")))
    };
    let mut out = vec![line(header, true)];
    out.push(dim(&widths.iter().map(|w| "─".repeat(*w)).collect::<Vec<_>>().join("─┼─")));
    for row in rows { out.push(line(row, false)); }
    out
}
